### ISSUE TITLE
Enable sending data in connection message

### ISSUE THREAD

**Issue by Kampfgnom** [hapijs/nes#200]
Hi everyone. First things first: Sorry for "misusing" GitHub issues for something that is partly a support request.

# Background

In our infrastructure we are responding resources differently depending on some HTTP headers like e.g. `accept-language`.

We are on our way of providing many resources via Nes subscriptions in addition to HTTP and need to somehow know pieces of information that we until now got from HTTP headers.

# Possible solutions

For now we came up with some possible solutions and I would love to start a discussion on how you would approach the issue.

## Solution 1: Seal more info in the AUTH request cookie

It should be possible to just use the information from the initial AUTH request and either put it into the same cookie or somehow enrich the resulting socket with the headers of that request.

Here is the relevant code: https://github.com/hapijs/nes/blob/master/lib/index.js#L189

## Solution 2: Explicitly transfer additional information via a `message`

We could implement our own convention to send messages via [client.message()](https://github.com/hapijs/nes/blob/master/API.md#clientmessagemessage-callback). I envision something like `{ command: 'setSocketHeaders', headers: { 'accept-language': '...' } }`

## Solution 3: Supply information via subscription path elements

It could be possible to subscribe to something like `/resource?language=de`?

# Wrap up

I will for now refrain from summarizing which of the solutions I personally prefer and the pros and cons I think I could find, as I am only starting to investigate and don't want to influence any discussion too much.

Thanks for any opinion on this :)

**Comment by hueniverse** [comment]
Since this information doesn't change, I would put it an initial message or in the authentication step and then record this information with the socket object.

**Comment by Kampfgnom** [comment]
Thank you for your answer 👍  I'll try to summarize my own findings:

I would like to keep the needed client side changes as few as possible, while at the same time not adding too much additional magic.

**Solution 1 (adding and providing more info to the auth request)** would be my preferred solution on first thought, but I could not see any way to do that without modifying Nes. Hence the opened issue. Is that correct?

It looks like I have to modify both the server-side route handler[1] and also somehow allow sending the information from the client[2]. It does not look trivial and would certainly change the current implementation a bit.

[1] https://github.com/hapijs/nes/blob/122f77c708ee901f12536463c633e926ee5eba9b/lib/index.js#L168-L171
[2] https://github.com/hapijs/nes/blob/122f77c708ee901f12536463c633e926ee5eba9b/lib/client.js#L474-L476

**Solution 2 (sending a message)** would be my alternative way to go. It would of course require no changes to Nes, but we would have to envision our own conventions on top of the Nes protocol. I am not really convinced that this is the right way to go.

If a lightweight change to Nes would allow everyone to get the feature I need, I would definitely prefer that.


I am of course willing to code a proposal, but currently do not have a good enough overview of the project to see "the best" approach.

If you think that you would like something like this in Nes, you could roughly sketch an approach you would like implemented and I would try to help out ;)

**Comment by Kampfgnom** [comment]
Thank you for your answer 👍  I'll try to summarize my own findings:

I would like to keep the needed client side changes as few as possible, while at the same time not adding too much additional magic.

**Solution 1 (adding and providing more info to the auth request)** would be my preferred solution on first thought, but I could not see any way to do that without modifying Nes. Hence the opened issue. Is that correct?

It looks like I have to modify both the server-side route handler[1] and also somehow allow sending the information from the client[2]. It does not look trivial and would certainly change the current implementation a bit.

[1] https://github.com/hapijs/nes/blob/122f77c708ee901f12536463c633e926ee5eba9b/lib/index.js#L168-L171
[2] https://github.com/hapijs/nes/blob/122f77c708ee901f12536463c633e926ee5eba9b/lib/client.js#L474-L476

**Solution 2 (sending a message)** would be my alternative way to go. It would of course require no changes to Nes, but we would have to envision our own conventions on top of the Nes protocol. I am not really convinced that this is the right way to go.

If a lightweight change to Nes would allow everyone to get the feature I need, I would definitely prefer that.


I am of course willing to code a proposal, but currently do not have a good enough overview of the project to see "the best" approach.

If you think that you would like something like this in Nes, you could roughly sketch an approach you would like implemented and I would try to help out ;)

**Comment by hueniverse** [comment]
You can send whatever auth headers you want in connect, so that's one way to pass more information. But it does require you to hack that into your server auth strategy. I think a better solution is to let you send random blob as part of connect. But if I do that, it will be for hapi v17 only...

**Comment by Kampfgnom** [comment]
Yeah I already guessed that we would run into the `hapi@17` constraint 😉  I can totally understand that. 

I want to get rid of these damn callbacks, too. (`await new Promise(resolve => socket.publish('/topic', {}, resolve))` anyone? 😄 )

I would love to be able to send arbitrary data during connect some day and migrate to it once we are at hapi@17.

For now I think I'll have to go with the message approach.

**Comment by mtharrison** [comment]
@Kampfgnom how did you get on with @hueniverse's suggestions. What's your feelings on this request now?

**Comment by Kampfgnom** [comment]
Thanks for coming back to this 👍 

We implemented a custom communication protocol via Nes messages.

For example if we want to communicate the `accept-language` of a client to the server, we let him send the following message:

```
{
    command: 'setHeaders',
    headers: {
        'accept-language': 'de'
    }
}
```

This information is then attached to the socket on the server-side (`class LiveUpdatesSession`).

Of course this works as expected and we did not encounter many issue until now.

*But* I think we will run into more and more issues over the next few sprints. This is due to our requirements to 1) add more and more information that way (due to other reasons that we sadly cannot really control) and 2) time-out the sessions after some time, when the client's access token expires.

Keeping track of and re-sending these things that normally go into HTTP headers will probably become an issue for us .

In addition to that, there will be a few more additional stakeholders, that will want to consume our API in the next months. I fear that communicating the custom message protocol (and making it clear when to *re*-send what) to all of them will be harder than it would have been to tell them: "Just send everything with the connection options"
